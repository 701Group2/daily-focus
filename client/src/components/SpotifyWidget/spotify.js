export const authEndpoint = "https://accounts.spotify.com/authorize";
// Replace with your app's client ID, redirect URI and desired scopes
const clientId = "0d4fc9fafa1949a7a9e3e810e12076f0";
const clientSecret = "7579323d4eab49d9a922d500a5f65c47";
// encode client ID and secret with Base 64
const encodedCredentials = window.btoa(clientId + ":" + clientSecret);

const redirectUri = "http://localhost:3000/home";

const scopes = [
    "user-read-currently-playing",
    "user-read-recently-played",
    "user-read-playback-state",
    "user-top-read",
    "user-modify-playback-state",
];

/**
 * Gets the authorization code from the URL
 * @returns code as a string
 */
const getAuthCodeFromResponse = () => {
    return window.location.search
        .substring(1)
        .split("&")
        .reduce((initial, item) => {
            var parts = item.split("=");
            initial[parts[0]] = decodeURIComponent(parts[1]);
            return initial;
        }, {});
};

/**
 * Gets the access token from the spotify API
 * @returns object containing the access_code, expires_in, and refresh_token
 */
export const getAuthToken = () => {
    const code = getAuthCodeFromResponse().code;

    var body = new URLSearchParams();
    body.append("grant_type", "authorization_code");
    body.append("code", code);
    body.append("redirect_uri", redirectUri);

    const requestOptions = {
        method: "POST",
        headers: {
            "Content-Type": "application/x-www-form-urlencoded",
            "Authorization": `Basic ${encodedCredentials}`,
        },
        body: body,
    };
    return fetch("https://accounts.spotify.com/api/token", requestOptions)
        .then((response) => {
            if (!response.ok) {
                return false;
            }
            return response.json();
        })
        .then((data) => {
            return data;
        });
};

/**
 * Refreshes the access token using the refresh token
 * @returns object containing access_token and expires_in
 */
export const getRefreshedToken = () => {
    const refreshToken = localStorage.getItem("spotify-refresh-token");

    var body = new URLSearchParams();
    body.append("grant_type", "refresh_token");
    body.append("refresh_token", refreshToken);

    const requestOptions = {
        method: "POST",
        headers: {
            "Content-Type": "application/x-www-form-urlencoded",
            "Authorization": `Basic ${encodedCredentials}`,
        },
        body: body,
    };
    return fetch("https://accounts.spotify.com/api/token", requestOptions)
        .then((response) => {
            if (!response.ok) {
                return false;
            }
            return response.json();
        })
        .then((data) => {
            return data;
        });
};

/**
 * sets the expiry time of the accessToken in local storage to five minutes before the token actually expires
 * @param {*} expiresIn how many seconds the token lasts before expiring
 * @returns time that the token expires, in milliseconds from epoch
 */
export const setExpiryTime = (expiresIn) => {
    const tenMinutes = 10 * 60 * 1000;
    const expiryTime = expiresIn * 1000 + Date.now() - tenMinutes;
    localStorage.setItem("spotify-token-expiration", expiryTime);
    return expiryTime;
};

export const accessUrl = `${authEndpoint}?client_id=${clientId}&redirect_uri=${redirectUri}&scope=${scopes.join(
    "%20"
)}&response_type=code&show_dialog=true`;
