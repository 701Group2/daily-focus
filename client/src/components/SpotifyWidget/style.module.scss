.spotify {
    display: flex;
    align-items: center;
    flex-direction: column;
    width: 340px;
    height: fit-content;
    background-color: #ffffff;
    border-radius: 25px;
    box-shadow: 0px 4px 16px 8px rgba(0, 0, 0, 0.1);
}

.spotifyTitle {
    display: flex;
    align-items: center;
    justify-content: center;
    background: #30a0f5;
    height: 48px;
    width: 100%;
    border-top-left-radius: 25px;
    border-top-right-radius: 25px;
}

.spotifyTitleText {
    color: white;
    font-size: 24px;
    font-family: "Ropa Sans";
}

.spotifyBody {
    width: calc(100% - 40px);
    padding: 30px 20px;
    display: flex;
    align-items: center;
    justify-content: space-around;
}

.songImage {
    width: 90px;
    height: 90px;
    background-color: #30a0f5;
}

.spotifyTrackInfo {
    padding-left: 10px;
    display: flex;
    flex-direction: column;
}

.spotifyTrackTitle {
    margin-bottom: 2px;
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
}

.spotifyArtistName {
    font-size: 0.8rem;
    color: grey;
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
}
