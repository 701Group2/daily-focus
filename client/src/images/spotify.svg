<svg width="392" height="250" viewBox="0 0 392 250" fill="none" xmlns="http://www.w3.org/2000/svg">
<g filter="url(#filter0_d)">
<rect x="26" y="26" width="340" height="198" rx="25" fill="white"/>
</g>
<path d="M26 51C26 37.1929 37.1929 26 51 26H341C354.807 26 366 37.1929 366 51V74H26V51Z" fill="#30A0F5"/>
<path d="M164.736 40.992C165.6 40.992 166.728 41.12 168.12 41.376L168.072 43.176C166.76 42.92 165.648 42.792 164.736 42.792C163.808 42.792 163.136 42.976 162.72 43.344C162.32 43.712 162.12 44.272 162.12 45.024V45.648C162.12 46.048 162.168 46.368 162.264 46.608C162.376 46.832 162.568 47.032 162.84 47.208C163.112 47.368 163.52 47.536 164.064 47.712L166.296 48.456C167.368 48.824 168.12 49.312 168.552 49.92C169 50.528 169.224 51.416 169.224 52.584V53.04C169.224 54.464 168.808 55.528 167.976 56.232C167.16 56.936 165.816 57.288 163.944 57.288C163.032 57.288 161.864 57.128 160.44 56.808L160.488 54.96C161.816 55.28 162.952 55.44 163.896 55.44C165.096 55.44 165.944 55.264 166.44 54.912C166.936 54.544 167.184 53.976 167.184 53.208V52.416C167.184 51.856 167.056 51.432 166.8 51.144C166.56 50.84 166.128 50.584 165.504 50.376L163.032 49.56C161.928 49.192 161.16 48.72 160.728 48.144C160.296 47.552 160.08 46.664 160.08 45.48V45.216C160.08 43.824 160.448 42.776 161.184 42.072C161.936 41.352 163.12 40.992 164.736 40.992ZM176.831 45C177.839 45 178.623 45.344 179.183 46.032C179.743 46.72 180.023 47.712 180.023 49.008V53.28C180.023 54.528 179.695 55.512 179.039 56.232C178.399 56.936 177.527 57.288 176.423 57.288C176.007 57.288 175.559 57.2 175.079 57.024C174.615 56.864 174.183 56.672 173.783 56.448V61.44H171.743V45.288H173.663V46.992C174.543 45.664 175.599 45 176.831 45ZM176.423 46.824C175.959 46.824 175.487 47 175.007 47.352C174.543 47.688 174.135 48.152 173.783 48.744V54.624C174.119 54.896 174.487 55.112 174.887 55.272C175.303 55.416 175.679 55.488 176.015 55.488C177.327 55.488 177.983 54.656 177.983 52.992V49.296C177.983 47.648 177.463 46.824 176.423 46.824ZM187.117 45C188.541 45 189.597 45.384 190.285 46.152C190.989 46.904 191.341 47.976 191.341 49.368V52.92C191.341 54.312 190.989 55.392 190.285 56.16C189.597 56.912 188.541 57.288 187.117 57.288C185.693 57.288 184.629 56.912 183.925 56.16C183.237 55.392 182.893 54.312 182.893 52.92V49.368C182.893 47.976 183.237 46.904 183.925 46.152C184.629 45.384 185.693 45 187.117 45ZM187.117 46.704C186.429 46.704 185.893 46.92 185.509 47.352C185.125 47.784 184.933 48.4 184.933 49.2V53.088C184.933 53.888 185.125 54.504 185.509 54.936C185.893 55.368 186.429 55.584 187.117 55.584C187.805 55.584 188.341 55.368 188.725 54.936C189.109 54.504 189.301 53.888 189.301 53.088V49.176C189.301 48.376 189.109 47.768 188.725 47.352C188.341 46.92 187.805 46.704 187.117 46.704ZM196.924 45.288H200.044V46.968H196.924V53.832C196.924 54.44 197.044 54.88 197.284 55.152C197.54 55.408 197.932 55.536 198.46 55.536C199.132 55.536 199.788 55.456 200.428 55.296L200.476 57C199.772 57.192 199.132 57.288 198.556 57.288C197.276 57.288 196.34 57.008 195.748 56.448C195.172 55.872 194.884 54.944 194.884 53.664V46.968H193.204V45.288H194.884L195.268 42.168H196.924V45.288ZM204.673 45.288V57H202.633V45.288H204.673ZM203.641 40.56C204.073 40.56 204.401 40.656 204.625 40.848C204.865 41.04 204.985 41.312 204.985 41.664V42.36C204.985 42.712 204.865 42.984 204.625 43.176C204.401 43.352 204.073 43.44 203.641 43.44C203.225 43.44 202.897 43.352 202.657 43.176C202.433 42.984 202.321 42.712 202.321 42.36V41.664C202.321 41.312 202.433 41.04 202.657 40.848C202.897 40.656 203.225 40.56 203.641 40.56ZM212.344 40.32C212.872 40.32 213.376 40.376 213.856 40.488V42.12C213.376 42.04 212.88 42 212.368 42C211.872 42 211.48 42.16 211.192 42.48C210.92 42.8 210.784 43.288 210.784 43.944V45.288H213.184V46.968H210.784V57H208.744V46.968H207.064V45.288H208.744V44.064C208.76 42.816 209.072 41.88 209.68 41.256C210.288 40.632 211.176 40.32 212.344 40.32ZM224.068 45.288L218.692 61.44H216.652L218.188 57H217.42L214.084 45.288H216.316L218.836 55.536L221.956 45.288H224.068Z" fill="white"/>
<rect x="45" y="104" width="90" height="90" fill="#30A0F5"/>
<path d="M163.302 106.994C163.95 106.994 164.796 107.09 165.84 107.282L165.804 108.632C164.82 108.44 163.986 108.344 163.302 108.344C162.606 108.344 162.102 108.482 161.79 108.758C161.49 109.034 161.34 109.454 161.34 110.018V110.486C161.34 110.786 161.376 111.026 161.448 111.206C161.532 111.374 161.676 111.524 161.88 111.656C162.084 111.776 162.39 111.902 162.798 112.034L164.472 112.592C165.276 112.868 165.84 113.234 166.164 113.69C166.5 114.146 166.668 114.812 166.668 115.688V116.03C166.668 117.098 166.356 117.896 165.732 118.424C165.12 118.952 164.112 119.216 162.708 119.216C162.024 119.216 161.148 119.096 160.08 118.856L160.116 117.47C161.112 117.71 161.964 117.83 162.672 117.83C163.572 117.83 164.208 117.698 164.58 117.434C164.952 117.158 165.138 116.732 165.138 116.156V115.562C165.138 115.142 165.042 114.824 164.85 114.608C164.67 114.38 164.346 114.188 163.878 114.032L162.024 113.42C161.196 113.144 160.62 112.79 160.296 112.358C159.972 111.914 159.81 111.248 159.81 110.36V110.162C159.81 109.118 160.086 108.332 160.638 107.804C161.202 107.264 162.09 106.994 163.302 106.994ZM171.545 110C172.613 110 173.405 110.288 173.921 110.864C174.449 111.428 174.713 112.232 174.713 113.276V115.94C174.713 116.984 174.449 117.794 173.921 118.37C173.405 118.934 172.613 119.216 171.545 119.216C170.477 119.216 169.679 118.934 169.151 118.37C168.635 117.794 168.377 116.984 168.377 115.94V113.276C168.377 112.232 168.635 111.428 169.151 110.864C169.679 110.288 170.477 110 171.545 110ZM171.545 111.278C171.029 111.278 170.627 111.44 170.339 111.764C170.051 112.088 169.907 112.55 169.907 113.15V116.066C169.907 116.666 170.051 117.128 170.339 117.452C170.627 117.776 171.029 117.938 171.545 117.938C172.061 117.938 172.463 117.776 172.751 117.452C173.039 117.128 173.183 116.666 173.183 116.066V113.132C173.183 112.532 173.039 112.076 172.751 111.764C172.463 111.44 172.061 111.278 171.545 111.278ZM180.881 110C181.649 110 182.237 110.264 182.645 110.792C183.065 111.308 183.275 112.046 183.275 113.006V119H181.745V113.222C181.745 112.022 181.355 111.422 180.575 111.422C180.227 111.422 179.873 111.554 179.513 111.818C179.153 112.082 178.841 112.442 178.577 112.898V119H177.047V110.216H178.487V111.53C178.811 111.038 179.171 110.66 179.567 110.396C179.975 110.132 180.413 110 180.881 110ZM188.681 110C189.209 110 189.689 110.078 190.121 110.234L192.929 110.198V111.548L191.453 111.512C191.657 111.944 191.759 112.394 191.759 112.862V113.258C191.759 114.134 191.489 114.842 190.949 115.382C190.409 115.91 189.653 116.174 188.681 116.174C188.057 116.174 187.547 116.114 187.151 115.994C186.887 116.21 186.755 116.456 186.755 116.732V116.84C186.755 117.032 186.815 117.176 186.935 117.272C187.055 117.356 187.289 117.41 187.637 117.434L190.319 117.632C190.871 117.68 191.315 117.89 191.651 118.262C191.999 118.622 192.173 119.078 192.173 119.63V120.296C192.173 120.812 192.071 121.232 191.867 121.556C191.663 121.88 191.285 122.126 190.733 122.294C190.193 122.462 189.413 122.546 188.393 122.546C188.009 122.546 187.553 122.504 187.025 122.42C186.509 122.336 186.017 122.228 185.549 122.096L185.603 120.674C186.719 121.046 187.667 121.232 188.447 121.232C189.095 121.232 189.575 121.19 189.887 121.106C190.211 121.034 190.421 120.92 190.517 120.764C190.625 120.608 190.679 120.374 190.679 120.062V119.918C190.679 119.642 190.601 119.426 190.445 119.27C190.301 119.114 190.091 119.03 189.815 119.018L187.115 118.838C186.539 118.802 186.101 118.64 185.801 118.352C185.501 118.064 185.351 117.674 185.351 117.182V116.822C185.351 116.558 185.417 116.3 185.549 116.048C185.693 115.796 185.897 115.574 186.161 115.382C185.957 115.154 185.807 114.866 185.711 114.518C185.627 114.158 185.585 113.72 185.585 113.204V112.808C185.585 111.956 185.849 111.278 186.377 110.774C186.905 110.258 187.673 110 188.681 110ZM188.681 111.278C188.177 111.278 187.787 111.41 187.511 111.674C187.247 111.926 187.115 112.292 187.115 112.772V113.366C187.115 113.81 187.151 114.14 187.223 114.356C187.307 114.56 187.451 114.704 187.655 114.788C187.859 114.86 188.165 114.896 188.573 114.896C189.125 114.896 189.539 114.764 189.815 114.5C190.091 114.236 190.229 113.858 190.229 113.366V112.826C190.229 112.322 190.157 111.884 190.013 111.512C189.869 111.452 189.671 111.398 189.419 111.35C189.179 111.302 188.933 111.278 188.681 111.278Z" fill="black"/>
<path d="M163.452 127.83L166.266 137H164.978L164.376 134.872H161.002L160.4 137H159.182L161.982 127.83H163.452ZM162.696 128.922L161.296 133.808H164.082L162.696 128.922ZM170.345 130C170.475 130 170.578 130.009 170.653 130.028V131.372C170.513 131.344 170.331 131.33 170.107 131.33C169.453 131.33 168.935 131.657 168.553 132.31V137H167.363V130.168H168.483V131.19C168.735 130.807 169.015 130.513 169.323 130.308C169.64 130.103 169.981 130 170.345 130ZM173.644 130.168H175.464V131.148H173.644V135.152C173.644 135.507 173.714 135.763 173.854 135.922C174.003 136.071 174.232 136.146 174.54 136.146C174.932 136.146 175.314 136.099 175.688 136.006L175.716 137C175.305 137.112 174.932 137.168 174.596 137.168C173.849 137.168 173.303 137.005 172.958 136.678C172.622 136.342 172.454 135.801 172.454 135.054V131.148H171.474V130.168H172.454L172.678 128.348H173.644V130.168ZM178.164 130.168V137H176.974V130.168H178.164ZM177.562 127.41C177.814 127.41 178.005 127.466 178.136 127.578C178.276 127.69 178.346 127.849 178.346 128.054V128.46C178.346 128.665 178.276 128.824 178.136 128.936C178.005 129.039 177.814 129.09 177.562 129.09C177.319 129.09 177.128 129.039 176.988 128.936C176.857 128.824 176.792 128.665 176.792 128.46V128.054C176.792 127.849 176.857 127.69 176.988 127.578C177.128 127.466 177.319 127.41 177.562 127.41ZM182.079 130C182.471 130 182.942 130.056 183.493 130.168L183.465 131.176C182.979 131.064 182.536 131.008 182.135 131.008C181.295 131.008 180.875 131.279 180.875 131.82V131.946C180.875 132.114 180.917 132.249 181.001 132.352C181.094 132.455 181.257 132.562 181.491 132.674L182.807 133.29C183.6 133.663 183.997 134.209 183.997 134.928V135.236C183.987 135.88 183.791 136.365 183.409 136.692C183.035 137.009 182.443 137.168 181.631 137.168C181.211 137.168 180.609 137.07 179.825 136.874L179.881 135.81C180.581 136.015 181.155 136.118 181.603 136.118C182.06 136.118 182.377 136.057 182.555 135.936C182.732 135.805 182.821 135.595 182.821 135.306V135.166C182.821 134.942 182.774 134.765 182.681 134.634C182.597 134.503 182.452 134.391 182.247 134.298L180.889 133.696C180.478 133.509 180.175 133.285 179.979 133.024C179.792 132.763 179.699 132.459 179.699 132.114V131.918C179.717 131.283 179.913 130.807 180.287 130.49C180.669 130.163 181.267 130 182.079 130ZM186.864 130.168H188.684V131.148H186.864V135.152C186.864 135.507 186.934 135.763 187.074 135.922C187.224 136.071 187.452 136.146 187.76 136.146C188.152 136.146 188.535 136.099 188.908 136.006L188.936 137C188.526 137.112 188.152 137.168 187.816 137.168C187.07 137.168 186.524 137.005 186.178 136.678C185.842 136.342 185.674 135.801 185.674 135.054V131.148H184.694V130.168H185.674L185.898 128.348H186.864V130.168Z" fill="#808080"/>
<path d="M258 171.5L236.25 184.057V158.943L258 171.5Z" fill="#6C6C6C"/>
<path d="M302 171.5L284.75 181.459V161.541L302 171.5Z" fill="#6C6C6C"/>
<path d="M320 171.5L302.75 181.459V161.541L320 171.5Z" fill="#6C6C6C"/>
<path d="M302 171.5L284.75 181.459V161.541L302 171.5Z" fill="#6C6C6C"/>
<path d="M320 171.5L302.75 181.459V161.541L320 171.5Z" fill="#6C6C6C"/>
<path d="M167 171.5L184.25 161.541V181.459L167 171.5Z" fill="#6C6C6C"/>
<path d="M185 171.5L202.25 161.541V181.459L185 171.5Z" fill="#6C6C6C"/>
<defs>
<filter id="filter0_d" x="0" y="0" width="392" height="250" filterUnits="userSpaceOnUse" color-interpolation-filters="sRGB">
<feFlood flood-opacity="0" result="BackgroundImageFix"/>
<feColorMatrix in="SourceAlpha" type="matrix" values="0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 127 0"/>
<feMorphology radius="9" operator="dilate" in="SourceAlpha" result="effect1_dropShadow"/>
<feOffset/>
<feGaussianBlur stdDeviation="8.5"/>
<feColorMatrix type="matrix" values="0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0.25 0"/>
<feBlend mode="normal" in2="BackgroundImageFix" result="effect1_dropShadow"/>
<feBlend mode="normal" in="SourceGraphic" in2="effect1_dropShadow" result="shape"/>
</filter>
</defs>
</svg>
