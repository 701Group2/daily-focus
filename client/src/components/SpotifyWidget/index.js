import React, { useState, useEffect } from "react";
import { makeStyles } from "@material-ui/core";
import Button from "@material-ui/core/Button";
import IconButton from "@material-ui/core/IconButton";
import style from "./style.module.scss";
import PlayArrowIcon from "@material-ui/icons/PlayArrow";
import FastForwardIcon from "@material-ui/icons/FastForward";
import FastRewindIcon from "@material-ui/icons/FastRewind";
import PauseIcon from "@material-ui/icons/Pause";
import { accessUrl, getAuthToken, getRefreshedToken, setExpiryTime } from "./spotify";
import SpotifyWebApi from "spotify-web-api-js";

const spotify = new SpotifyWebApi();

const useStyles = makeStyles({
    root: {
        margin: "30px 20px",
        background: "#1DB954",
        borderRadius: 50,
        padding: "4px 25px",
        width: "calc(100% - 40px)",
    },
    label: {
        textTransform: "capitalize",
        fontSize: "24px",
    },
});

export default function SpotifyWidget() {
    const classes = useStyles();
    const [play, setPlay] = useState(false);
    const [isLoggedIn, setIsLoggedIn] = useState(false);
    const [item, setItem] = useState(false);

    useEffect(() => {
        var _token = localStorage.getItem("spotify-token");

        if (!_token) {
            setIsLoggedIn(false);

            // the authorisation code comes from the query of page's url after the user logs in to spotify
            // check to see if the code is there. If it is, then use it get the access token
            if (window.location.search !== "") {
                getAuthToken().then((tokenData) => {
                    // store the tokens and expiry time in local storage
                    _token = tokenData.access_token;
                    const refreshToken = tokenData.refresh_token;

                    localStorage.setItem("spotify-token", _token);
                    localStorage.setItem("spotify-refresh-token", refreshToken);
                    setExpiryTime(tokenData.expires_in);

                    // the authorisation code is no longer needed, remove it from the URl
                    window.location.search = "";
                });
            }
        }

        if (_token) {
            var token_expiration = localStorage.getItem("spotify-token-expiration");
            localStorage.setItem("spotify-token-expiration", Date.now());

            // if the token is expired, try to refresh the access token
            if (token_expiration && token_expiration <= Date.now()) {
                getRefreshedToken().then((refreshData) => {
                    if (refreshData) {
                        // set the new access token
                        _token = refreshData.access_token;
                        onSuccessfulToken(_token);
                    } else {
                        // if the token could not be refreshed, the user needs to log in again
                        _token = null;
                        setIsLoggedIn(false);
                    }
                });
            } else {
                onSuccessfulToken(_token);
            }
        }
    }, []);

    /**
     * Run after the spotify access token is confirmed to be valid
     * @param {*} token the access token
     */
    const onSuccessfulToken = (token) => {
        spotify.setAccessToken(token);
        setIsLoggedIn(true);

        //not needed but just logging to show user info
        spotify.getMe().then((user) => {
            console.log(user);
        });

        spotify.getMyCurrentPlaybackState().then((track) => {
            setPlay(track.is_playing);
        });
    };

    const handlePlay = () => {
        spotify.getMyDevices().then((devices) => {
            if (play) {
                spotify.pause();
            } else {
                spotify
                    .play({
                        device_id: devices.devices[0]?.id,
                        context_uri: "spotify:playlist:0vvXsWCC9xrXsKd4FyS8kM",
                    })
                    .then(() => {
                        spotify.getMyCurrentPlayingTrack().then((r) => {
                            setItem(r.item);
                        });
                    });
            }
        });
    };

    const skipNext = () => {
        spotify.skipToNext().then(() => {
            spotify.getMyCurrentPlayingTrack().then((r) => {
                setItem(r.item);
                setPlay(true);
            });
        });
    };

    const skipPrevious = () => {
        spotify.skipToPrevious().then(() => {
            spotify.getMyCurrentPlayingTrack().then((r) => {
                setItem(r.item);
                setPlay(true);
            });
        });
    };

    return (
        <div className={style.spotify}>
            <div className={style.spotifyTitle}>
                <div className={style.spotifyTitleText}>Spotify</div>
            </div>
            {!isLoggedIn ? (
                <Button
                    href={accessUrl}
                    variant="contained"
                    color="primary"
                    classes={{ root: classes.root, label: classes.label }}
                >
                    Login with Spotify
                </Button>
            ) : (
                <div className={style.spotifyBody}>
                    <img
                        className={style.songImage}
                        src={item?.album?.images[0]?.url}
                        alt={item?.album?.name}
                    />
                    <div>
                        <IconButton aria-label="rewind" onClick={skipPrevious}>
                            <FastRewindIcon fontSize="large" />
                        </IconButton>
                        <IconButton
                            aria-label="play"
                            onClick={() => {
                                setPlay(!play);
                                handlePlay();
                            }}
                        >
                            {play ? (
                                <PauseIcon fontSize="large" />
                            ) : (
                                <PlayArrowIcon fontSize="large" />
                            )}
                        </IconButton>
                        <IconButton aria-label="fastforward" onClick={skipNext}>
                            <FastForwardIcon fontSize="large" />
                        </IconButton>
                    </div>
                </div>
            )}
        </div>
    );
}
