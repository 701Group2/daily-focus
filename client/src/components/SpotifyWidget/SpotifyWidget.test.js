import React from "react";
import { shallow } from "enzyme";
import ShallowRenderer from "react-test-renderer/shallow";
import Button from "@material-ui/core/Button";

import SpotifyWidget from "./";
import { accessUrl } from "./spotify";

let component;

beforeEach(() => {
    component = shallow(<SpotifyWidget />);
});

test("it renders correctly", () => {
    const shallowRenderer = new ShallowRenderer();
    const snapshotComponent = shallowRenderer.render(<SpotifyWidget />);
    expect(snapshotComponent).toMatchSnapshot();
});

test("it contains Material UI Button component", () => {
    expect(component.find(Button)).toHaveLength(1);
    expect(component.find(Button).props().color).toEqual("primary");
});

test("it contains spotify container", () => {
    expect(component.find(".spotify")).toHaveLength(1);
});

test("it contains spotify container", () => {
    expect(component.find(".spotify")).toHaveLength(1);
});

test("it contains spotify title", () => {
    expect(component.find(".spotifyTitle")).toHaveLength(1);
    expect(component.find(".spotifyTitleText")).toHaveLength(1);
    expect(component.find(".spotifyTitleText").props().children).toEqual("Spotify");
});

test("the spotify button has the correct text", () => {
    expect(component.find(Button).props().children).toEqual("Login with Spotify");
});

test("the spotify button leads to a new landing page", () => {
    expect(component.find(Button).props().href).toEqual(accessUrl);
    expect(component.find(Button).props().variant).toEqual("contained");
});
