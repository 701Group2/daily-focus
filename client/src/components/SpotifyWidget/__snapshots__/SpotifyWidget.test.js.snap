// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders correctly 1`] = `
<div
  className="spotify"
>
  <div
    className="spotifyTitle"
  >
    <div
      className="spotifyTitleText"
    >
      Spotify
    </div>
  </div>
  <WithStyles(ForwardRef(Button))
    classes={
      Object {
        "label": "makeStyles-label-2",
        "root": "makeStyles-root-1",
      }
    }
    color="primary"
    href="https://accounts.spotify.com/authorize?client_id=0d4fc9fafa1949a7a9e3e810e12076f0&redirect_uri=http://localhost:3000/home&scope=user-read-currently-playing%20user-read-recently-played%20user-read-playback-state%20user-top-read%20user-modify-playback-state&response_type=code&show_dialog=true"
    variant="contained"
  >
    Login with Spotify
  </WithStyles(ForwardRef(Button))>
</div>
`;
